### PR TITLE
sequence parallel for uneven heads

### PR DESCRIPTION
**Pull request by inkcherry**
In sequence_parallel (Ulysses), the sequence parallel size is constrained by the requirement to be divisible by the number of heads, which prevents some models/workloads from setting a specific sequence parallel size. This PR implements uneven all-to-all heads splitting. 

- both support  batch first (b,s,...) and seq_len first(s,b..) layout.
- Added unit tests with numerical checks.     Locally also tested with **7 heads with sp=4** and **20 heads with sp=8**, and it passed.

### CODE REVIEW

File: deepspeed/sequence/layer.py [hunk 1]
```diff
@@ -38,8 +40,69 @@ def post_func(input):
...
+        dist.all_to_all_single(output,input,output_split_sizes=output_splits,\
+            input_split_sizes=input_splits,group=group)
+
+        ###[seq_ws*local_heads, ..] to [seq_ws, local_heads, ..]
+        output = output.view(seq_world_size, local_heads, *output.shape[1:])
+        ###[seq_ws,local_heads,b,seq_len,..] to [seq_ws,seq_len,b,local_heads,..]
+        output = output.transpose(1, 3).contiguous()
+        ###[seq_ws*local_seq_len, b, local_heads,...]
+        output = output.view(inp_shape[gather_idx] * seq_world_size, *output.shape[2:]).contiguous()
+    if scatter_idx < 2:
```
Comment (delock) [inline deepspeed/sequence/layer.py:64]: This line should be `else`

File: deepspeed/sequence/layer.py [hunk 2]
```diff
@@ -38,8 +40,69 @@ def post_func(input):
...
+
+
 def single_all_to_all(input, scatter_idx, gather_idx, batch_dim_idx, group, async_op=False, handle=None, type=None):
     seq_world_size = dist.get_world_size(group)
+
+    num_heads = input.shape[2]
+
+    if get_num_kv_heads() is not None or num_heads % seq_world_size != 0:
+        # assume here that the number of heads for q is consistent with kv
+        # or require additional logic
```
Comment (delock) [inline deepspeed/sequence/layer.py:98]: Should specify in comments that 'GQA are not supported' to help understanding.

File: deepspeed/sequence/layer.py (hunk 1 above)
Comment (inkcherry) [reply deepspeed/sequence/layer.py]: Thanks for pointing that out, changed

File: deepspeed/sequence/layer.py (hunk 2 above)
Comment (inkcherry) [reply deepspeed/sequence/layer.py]: yes, added

File: deepspeed/sequence/layer.py [hunk 3]
```diff
@@ -38,8 +40,130 @@ def post_func(input):
...
+                                 *output.shape[3:]).contiguous()
+        elif batch_dim_idx == 1:
+            output = output.transpose(1, 3).contiguous()
+            ###[seq_ws*local_seq_len, b, local_heads,...]
+            output = output.view(inp_shape[gather_idx] * seq_world_size, *output.shape[2:]).contiguous()
+    else:
+        # The compatibility handling of 4D and 3D tensors, standardizing to 3D.
+        input = input.view(input.shape[0], input.shape[1], -1)
+
+        if batch_dim_idx == 0:  #b,s,h
```
Comment (delock) [inline deepspeed/sequence/layer.py:76]: Should assert for batch_dim_idex != 0 or 1

Comment (inkcherry) [reply deepspeed/sequence/layer.py:78]: added.

**Comment by delock** [comment]
@inkcherry thanks, I have no further questions.  Hi @tjruwase @loadams this PR is to enable sequence parallel for model with number of heads not power of two, which is requested from customer.  Can this PR be reviewed?  Thanks!

**Comment by loadams** [comment]
> @inkcherry thanks, I have no further questions. Hi @tjruwase @loadams this PR is to enable sequence parallel for model with number of heads not power of two, which is requested from customer. Can this PR be reviewed? Thanks!

Thanks @delock, things look good now, but we will just need to get a review.

**Review by tjruwase — APPROVED** [review]

File: deepspeed/sequence/layer.py [hunk 4]
```diff
@@ -38,8 +40,132 @@ def post_func(input):
...
+        total_h = int(inp_shape[gather_idx] * heads_scale_coeff)
+        output = torch.empty(output_buff_d1_size, input.shape[1], device=input.device, dtype=input.dtype)
+        dist.all_to_all_single(output,input,output_split_sizes=output_splits, \
+            input_split_sizes=input_splits,group=group)
+        ##################
+        #suppose 7 heads divide into 4 ranks [2,2,2,1]
+        #chunk_num_heads_small=floor(7/4)=1
+        #chunk_num_heads_large=ceil(7/4)=2
+        #num_chunk_heads_large=len([2,2,2])=3, all2all_buffer_counts
+        #num_chunk_heads_large=len([1])=1, all2all_buffer_counts
```
Comment (tjruwase) [inline deepspeed/sequence/layer.py:107]:
```suggestion
        #num_chunk_heads_small=len([1])=1, all2all_buffer_counts
```

Comment (tjruwase) [reply deepspeed/sequence/layer.py:107]: @inkcherry, what do you think of this correction to the comment?

Comment (inkcherry) [reply deepspeed/sequence/layer.py]: yes, thanks for point that out.

**Pull request merged**